File: store/src/test/java/com/nytimes/android/external/store3/SequentialTest.java
package com.nytimes.android.external.store3;

import com.nytimes.android.external.store.util.Result;
import com.nytimes.android.external.store3.base.impl.BarCode;
import com.nytimes.android.external.store3.base.impl.Store;
import com.nytimes.android.external.store3.base.impl.StoreBuilder;
import io.reactivex.Single;
import org.junit.Before;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class SequentialTest {

    int networkCalls = 0;
    private Store<Integer, BarCode> store;

    @Before
    public void setUp() {
        networkCalls = 0;
        store = StoreBuilder.<Integer>barcode()
                .fetcher(barcode -> Single.fromCallable(() -> networkCalls++))
                .open();
    }

    @Test
    public void sequentially() {
        BarCode b = new BarCode("one", "two");
        store.get(b).test().awaitTerminalEvent();
        store.get(b).test().awaitTerminalEvent();

        assertThat(networkCalls).isEqualTo(1);
    }

    @Test
    public void sequentiallyWithResult() {
        BarCode b = new BarCode("one", "two");
        store.getWithResult(b).test().awaitTerminalEvent();
        store.getWithResult(b).test().awaitTerminalEvent();

        assertThat(networkCalls).isEqualTo(1);
    }

    @Test
    public void parallel() {
        BarCode b = new BarCode("one", "two");
        Single<Integer> first = store.get(b);
        Single<Integer> second = store.get(b);

        first.test().awaitTerminalEvent();
        second.test().awaitTerminalEvent();

        assertThat(networkCalls).isEqualTo(1);
    }

    @Test
    public void parallelWithResult() {
        BarCode b = new BarCode("one", "two");
        Single<Result<Integer>> first = store.getWithResult(b);
        Single<Result<Integer>> second = store.getWithResult(b);

        first.test().awaitTerminalEvent();
        second.test().awaitTerminalEvent();

        assertThat(networkCalls).isEqualTo(1);
    }
}

File: store/src/test/java/com/nytimes/android/external/store3/SequentialTest.kt
package com.nytimes.android.external.store3

import com.nytimes.android.external.store3.base.impl.BarCode
import com.nytimes.android.external.store3.base.impl.StoreBuilder
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class SequentialTest {

    var networkCalls = 0
    private var store = StoreBuilder.barcode<Int>()
            .fetcher { networkCalls++ }
            .open()

    @Test
    fun sequentially() = runBlocking<Unit> {
        val b = BarCode("one", "two")
        store.get(b)
        store.get(b)

        assertThat(networkCalls).isEqualTo(1)
    }

//    @Test
//    fun sequentiallyWithResult() = runBlocking<Unit> {
//        val b = BarCode("one", "two")
//        store.getWithResult(b)
//        store.getWithResult(b)
//
//        assertThat(networkCalls).isEqualTo(1)
//    }

    @Test
    fun parallel() = runBlocking<Unit> {
        val b = BarCode("one", "two")
        val deferred = async { store.get(b) }
        store.get(b)
        deferred.await()

        assertThat(networkCalls).isEqualTo(1)
    }

//    @Test
//    fun parallelWithResult() = runBlocking<Unit> {
//        val b = BarCode("one", "two")
//        val first = store.getWithResult(b)
//        val second = store.getWithResult(b)
//
//        assertThat(networkCalls).isEqualTo(1)
//    }
}
